### PR TITLE
Implement I2C Module's APIs: write, writeByte, writeBlock, on('data')…

### PR DESCRIPTION
**Pull request by nova0821**
…, stream

- Implement I2C APIs for Linux platform (write, writeByte, writeBlock, on, stream)
- Tested on Raspberry Pi
- Add Testcase for I2C APIs related read (read, readByte, readBytes, on('data), stream, scan)

IoT.js-DCO-1.0-Signed-off-by: Jaechul Yang jc08.yang@samsung.com

### CODE REVIEW

**Comment by nova0821** [comment]
https://github.com/Samsung/iotjs/pull/450 is landed, so I will rebase it.

**Comment by zherczeg** [comment]
LGTM

**Comment by nova0821** [comment]
@zherczeg write, writeByte Testcase is added. Please review again.

File: test/run_pass/test_i2c.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  @TIMEOUT=10
```
Comment (LaszloLango) [inline test/run_pass/test_i2c.js]: are these metainfos needed after #450 ?

**Comment by LaszloLango** [comment]
LGTM, otherwise

File: src/module/iotjs_module_i2c.cpp [hunk 2]
```diff
@@ -49,6 +48,19 @@ I2c* I2c::GetInstance() {
 }
 
 
+void GetI2cArray(JObject* jarray, I2cReqData* req_data) {
+  // FIXME: Need to implement a function to get array info from JObject Array.
+  req_data->buf_len = jarray->GetProperty("length").GetNumber();
```
Comment (nova0821) [inline src/module/iotjs_module_i2c.cpp]: There is function ``IsArray`` to check it is array object or not. I checked the type of JObject before calling this function.

File: test/run_pass/test_i2c.js (hunk 1 above)
Comment (nova0821) [inline test/run_pass/test_i2c.js]: I'll remove it. Thanks for review.

File: src/module/iotjs_module_i2c.cpp (hunk 2 above)
Comment (nova0821) [inline src/module/iotjs_module_i2c.cpp]: For more safety, I'll follow your comment. I'll add assert for check there is length property or not.

Comment (nova0821) [inline src/module/iotjs_module_i2c.cpp]: Fixed.

File: src/js/i2c.js [hunk 3]
```diff
@@ -199,13 +193,14 @@ I2C.prototype.stream = function(cmd, len, delay) {
...
       } else {
-        return _this.emit('data', {
+        _this.emit('data', {
           address: _this.address,
           data: data,
           cmd: cmd,
           length: len,
           timestamp: Date.now()
         });
+        _this.stream(cmd, len, delay);
```
Comment (chokobole) [inline src/js/i2c.js]: Originally, this callback function returns boolean which says event emitter emits event successfully or not. And I think the correspondency with the past is broken.

File: src/module/iotjs_module_i2c.h [hunk 4]
```diff
@@ -41,15 +41,16 @@ enum I2cError {
...
 struct I2cReqData {
   String device;
   char* buf_data;
-  int buf_len;
-  int8_t byte;
-  int8_t cmd;
+  uint8_t buf_len;
+  uint8_t byte;
+  uint8_t cmd;
   int32_t delay;
```
Comment (chokobole) [inline src/module/iotjs_module_i2c.h]: why don't type delay to ``uint32_t``?

Comment (nova0821) [inline src/module/iotjs_module_i2c.h]: For negative delay. For example, if you pass negative number to ``setTimeout``, it is changed to zero. Also, I think ``int32_t`` has enough range. (-2^31 ~ 2^31)

**Comment by zherczeg** [comment]
still LGTM

**Pull request merged**